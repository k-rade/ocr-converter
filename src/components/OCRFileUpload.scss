.ocr-file-upload {
  display: flex;
  align-items: center;
  justify-content: center;
  flex-direction: column;
  width: 100%;
  padding: 0 1rem;
  height: 16rem;
  position: relative;
  border: 2px dashed var(--color-gray-600);
  background-color: var(--color-gray-200);



  &-container {
    display: flex;
    flex-direction: column;
    align-items: center;
    text-align: center;

    &-txt {
      &-title {
        font-size: 1.25rem;
        color: var(--color-white-one);
        margin: 0.625rem 0;
      }

      &-desc {
        color: var(--color-gray-900);
        margin: 0;
        line-height: 1.35;
        font-size: 0.875rem;
      }
    }
  }

  .icon-file-upload,
  .icon-file-upload-success {
    width: 3rem;
    height: 3rem;
    fill: var(--color-white-one);
  }

  &-input {
    cursor: pointer;
    opacity: 0;
    position: absolute;
    height: 100%;
    width: 100%;

    &:disabled {
      cursor: not-allowed
    }
  }
}