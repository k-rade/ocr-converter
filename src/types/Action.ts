export enum ActionType {
  SET_IMAGE_FILE,
  SET_IMAGE_URL,
  SET_LANGUAGE,
  SET_PROGRESS,
  SET_ERROR,
  CONVERT_IMAGE_INIT,
  CONVERT_IMAGE_SUCCESS,
  CONVERT_IMAGE_ERROR,
  INIT_NEW_CONVERSION,
}

export interface Action {
  type: ActionType;
  payload?: any;
}
