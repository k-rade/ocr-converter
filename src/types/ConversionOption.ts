export enum ConversionOption {
  FILE_UPLOAD,
  BY_LINK,
}
