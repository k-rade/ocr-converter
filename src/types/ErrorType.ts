export enum ErrorType {
  NO_ERROR,
  CONVERSION_FAILED,
  UNSUPPORTED_FILE_FORMAT,
  MAX_FILE_SIZE_EXCEEDED,
}
