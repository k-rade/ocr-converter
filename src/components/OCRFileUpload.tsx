import { useState } from "react";
import { ErrorType } from "../types/ErrorType";
import { RiImageAddLine } from "react-icons/ri";

interface Props {
  setImage: (value: string) => void;
  handleError: (error: ErrorType) => void;
  disabled: boolean;
}

const SUPPORTED_FILE_FORMATS = [".jpg", "jpeg", ".png", ".bmp", ".pbm"];

const OCRFileUpload = ({
  setImage,
  handleError,
  disabled,
}: Props): JSX.Element => {
  const [file, setFile] = useState({ name: "", size: "" });

  const handleFileUpload = (e: any): void => {
    const file = e?.target?.files[0];

    if (file) {
      const isUnsupportedFileFormat = !SUPPORTED_FILE_FORMATS.some(
        (fileFormat) => fileFormat === file?.name.slice(-4)
      );

      if (isUnsupportedFileFormat) {
        handleError(ErrorType.UNSUPPORTED_FILE_FORMAT);
        return;
      }

      const isLargerThanFiveMegabytes = file?.size / 1000 > 5120;

      if (isLargerThanFiveMegabytes) {
        handleError(ErrorType.MAX_FILE_SIZE_EXCEEDED);
        return;
      }

      setFile({
        name: file?.name,
        size: `${(file?.size / 1000).toFixed(2)} KB`,
      });
      setImage(URL.createObjectURL(e.target?.files![0]));
    }
  };

  return (
    <div className="cursor-pointer flex border-dashed p-6 h-20">
      <input
        id="file-upload"
        disabled={disabled}
        className="sr-only"
        type="file"
        onChange={handleFileUpload}
        data-testid="ocr-file-upload-input"
      />
      <label
        htmlFor="file-upload"
        className="cursor-pointer flex justify-between w-full items-center gap-6"
      >
        <div>
          {file.name === "" ? (
            <div className="">
              <span className="text-white text-lg font-medium">
                Upload an image
              </span>{" "}
              <p className="text-neutral-500 text-sm">
                {file.size ? (
                  file.size
                ) : (
                  <>JPG, JPEG, PNG, BMP, PBM up to 5MB</>
                )}
              </p>
            </div>
          ) : (
            file.name
          )}
        </div>
        <RiImageAddLine size="2.5rem" className="fill-neutral-600" />
      </label>
    </div>
  );
};

export default OCRFileUpload;
