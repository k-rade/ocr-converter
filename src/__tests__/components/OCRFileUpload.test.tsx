import { act, render, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import OCRFileUpload from "../../components/OCRFileUpload";
import { ErrorType } from "../../types/ErrorType";

describe("OCRFileUpload", (): void => {
  beforeEach(() => {
    global.URL.createObjectURL = jest.fn();
  });

  it("should render", (): void => {
    const { baseElement } = render(
      <OCRFileUpload setImage={jest.fn} handleError={jest.fn} />
    );

    expect(baseElement).toMatchSnapshot();
  });

  it("should upload image", async (): Promise<void> => {
    const mockSetImage = jest.fn();
    const { getByTestId } = render(
      <OCRFileUpload setImage={mockSetImage} handleError={jest.fn} />
    );

    const fileInput = getByTestId("ocr-file-upload-input");
    const mockFile = new File(["hello"], "hello.png", { type: "image/png" });

    await waitFor(() => {
      userEvent.upload(fileInput, mockFile);
    });

    expect(mockSetImage).toHaveBeenCalled();
  });

  it("should handle error when file is unsupported format", async (): Promise<void> => {
    const mockHandleError = jest.fn();
    const { getByTestId } = render(
      <OCRFileUpload setImage={jest.fn} handleError={mockHandleError} />
    );

    const fileInput = getByTestId("ocr-file-upload-input");
    const mockFile = new File(["hello"], "hello.txt", { type: "text" });

    await waitFor(() => {
      userEvent.upload(fileInput, mockFile);
    });

    expect(mockHandleError).toHaveBeenCalledWith(
      ErrorType.UNSUPPORTED_FILE_FORMAT
    );
  });

  it("should handle error when maximum file size is exceeded", async (): Promise<void> => {
    const mockHandleError = jest.fn();
    const { getByTestId } = render(
      <OCRFileUpload setImage={jest.fn} handleError={mockHandleError} />
    );

    const FileInput = getByTestId("ocr-file-upload-input");
    const file = new File(["hello"], "hello.png", { type: "image/png" });
    Object.defineProperty(file, "size", { value: 10_000 * 10_000 + 1 });

    await waitFor(() => {
      userEvent.upload(FileInput, file);
    });

    expect(mockHandleError).toHaveBeenNthCalledWith(
      1,
      ErrorType.MAX_FILE_SIZE_EXCEEDED
    );
  });
});
