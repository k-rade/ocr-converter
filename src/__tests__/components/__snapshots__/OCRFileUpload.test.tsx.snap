// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OCRFileUpload should render 1`] = `
<body>
  <div>
    <div
      class="cursor-pointer flex p-5 h-20"
    >
      <input
        class="sr-only"
        data-testid="ocr-file-upload-input"
        disabled=""
        id="file-upload"
        type="file"
      />
      <label
        class="cursor-pointer flex justify-between w-full items-center gap-6"
        for="file-upload"
      >
        <div
          class="w-full max-w-lg truncate"
        >
          <div
            class="text-white text-lg font-medium truncate"
          >
            Upload an image
          </div>
          <div
            class="text-neutral-500 text-sm mt-0.5"
          >
             JPG, JPEG, PNG, BMP, PBM up to 5MB
          </div>
        </div>
        <div
          class="bg-neutral-700 p-2 rounded hover:bg-emerald-800"
        >
          <svg
            class="stroke-white"
            fill="none"
            height="1.5rem"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="1.5rem"
            xmlns="http://www.w3.org/2000/svg"
          >
            <line
              x1="12"
              x2="12"
              y1="5"
              y2="19"
            />
            <line
              x1="5"
              x2="19"
              y1="12"
              y2="12"
            />
          </svg>
        </div>
      </label>
    </div>
  </div>
</body>
`;
